### ISSUE TITLE
ParameterInterpolation.ipynb

### ISSUE THREAD

**Issue by sabaronett** [sabaronett/reboundx#22]
- [x] Fork from Dan's `MassEvolution.ipynb`
- [x] Use PyMesaReader to import stellar mass and radius data
- [x] 4 Myr simulation 8cf22e195112842b2cfd97adf3037544b59de758
- [x] Plot: _a_<sub>&oplus;</sub> & _R_<sub>&odot;</sub> vs. _t_, and  _M_<sub>&odot;</sub> vs. _t_
- [x] Debug `nan` interpolate issue with simple text import

> Other than that, it might be nice to take your guys' use case (updating stellar mass, radius, parameters) as an example (we can get rid of MassEvolution or rename it to ParameterInterpolation), and after editing the documentation I think we're done.

**Comment by sabaronett** [comment]
Hi @dtamayo,

I've committed a first draft of our use case as an iPython example in 8cf22e195112842b2cfd97adf3037544b59de758. Couple of questions:
- I noticed the `OrbitPlot`s look a little weird when there's just the star and a single particle (see [ParameterInterpolation.ipynb](https://github.com/sabaronett/reboundx/blob/8cf22e195112842b2cfd97adf3037544b59de758/ipython_examples/ParameterInterpolation.ipynb)). I've confirmed the same behavior in our previous [MassEvolution.ipynb](https://github.com/sabaronett/reboundx/blob/8cf22e195112842b2cfd97adf3037544b59de758/ipython_examples/MassEvolution.ipynb) when commenting out the outer two planets. Have you seen this before? Anyway to fix it without adding additional particles?
- As you can see, it takes about 11.5 minutes to run the 4 Myr sim, calling `.interpolate()` twice each loop. I've found in some other tests that this double-call roughly doubles the calculation time, i.e., if I only called `starmass.interpolate(rebx, t=T0+sim.t)` by itself it would only take about 6 minutes. How do these performance numbers sound to you? Should I do some more testing on this?
- I've used [py_mesa_reader](https://github.com/wmwolf/py_mesa_reader) for this initial version, but I'm assuming we want a method that's more "out-of-the-box," correct? Should the "reading in from a text file" be something we want to implement in REBOUNDx, similar to the new `Interpolator` class, or just a small Python script for use only in this example?

Let me know your thoughts. Thanks.

**Comment by dtamayo** [comment]
Hey @sabaronett ,

Great!

* Thanks a lot for catching that bug. We will sort that out in the main REBOUND repo, but for now you can get a reasonable plot by getting back both the fig and ax and setting the limits manually:

```
fig, ax = rebound.OrbitPlot(sim)
ax.set_ylim([-1,1])
```

* I think the computation of the interpolation should be completely negligible. The comparison should be to an integration run for the same amount of time but without updating the mass or radius. Also, we want to keep these examples to run in less than ~20 sec or so so people can run them easily and they pass quickly though the automated testing suite. You can do this easily by using

```
sim.integrator="whfast"
sim.dt = 0.1*sim.particles[1].P
```

Typically you should use a slightly smaller timestep than 10% of the innermost orbital period, but for a single planet case that's totally fine. That runs 4 Myr in about 10 sec on my laptop.

* Yeah I think we should leave the data generation/reading to the user, and just show how to incorporate it. So I would save your mesa output to text files 'm.txt' and 'r.txt', load those, show they're simple arrays, and then use them.

**Comment by sabaronett** [comment]
Hi Dan,

Thanks for getting back to me on everything.
- Should I go ahead and close #20 then?
- Since we're checking in this test whether the Sun will engulf the Earth, should I track/plot Earth's perihelion instead of semi-major axis? If so, is that just the same thing as `sim.particles[1].omega`?
- I've incorporated your suggestions (WHFast, larger timestep, controlled test w/o interpolations) in af112d6acfe3e7e5f5278a1bbf9d99d599346c98. Although you're correct&dash;the interpolation adds negligible time&dash;the 4 Myr runs take ~ 52 - 54 seconds. Any reason why it's taking me about 5 times longer than you? Could using Jupyter Lab (instead of the classic Notebook) make a difference? Here's my laptop's relevant specs:
   - Intel® Core™ i5 8265U: 4 cores (8 threads), max. Turbo Boost freq. up to 3.9 GHz
   - 8 GB DDR4-2400 SDRAM
- I'll work on splitting the input data files and adapting @Nofe4108's Python reader script into a short function for this notebook.

**Comment by sabaronett** [comment]
Hey @dtamayo, wanted to give you an update on reading the individual mass/radius text files. As of 444f4c2629f155a49f7bb054c414c139c5459439, I've uploaded an `m.txt` and `r.txt` with only two-columns each: one for times, and the other for the parameter values. I also removed over a thousand unnecessary entries of the data for times before and after our simulation window.

The issue I'm experiencing is that calls to `.interpolate(rebx, t=[time])` are returning `nan` for a reason that I cannot figure out. To demonstrate the issue, I've created two new&mdash;but nearly identical&mdash;notebooks (with all markdown removed) as part of this commit. If you compare [ParameterInterpolation-MesaReader.ipynb ](https://github.com/sabaronett/reboundx/blob/444f4c2629f155a49f7bb054c414c139c5459439/ipython_examples/ParameterInterpolation-MesaReader.ipynb) and [ParameterInterpolation-SimpleText.ipynb](https://github.com/sabaronett/reboundx/blob/444f4c2629f155a49f7bb054c414c139c5459439/ipython_examples/ParameterInterpolation-SimpleText.ipynb) side-by-side, you'll see that the `times`, `masses`, and `radii` arrays are of the same data type in both, differing only in size because of the removed lines. Yet, in cell [7] in both NBs, you'll see how a test call of `interpolate()` (at `t=` the first time entry in the arrays) works in the former, but returns `nan` in the latter. As you can also see in the latter NB, the subsequent simulation loop runs into the same issue, yielding invalid results and a longer computation time.

Any idea what might be causing this? Or any other testing suggestions?

**Comment by dtamayo** [comment]
Hey Stanley,

* No, that's not the pericenter. I think the more straightforward way is to set ``sim.collision = "direct"``, which will raise an exception when any particles collide.

* No those differences shouldn't matter at all. I'm also a bit confused why that's the case. Probably not worth worrying about right now

* I think the most logical way to organize it is if your example notebook just loads your txt files (taking them as given), and you create a separate repo that you can link to in your paper that shows people how to read a mesa output file and write out a text file. If we added the mesa reader into REBOUNDx and the example then for example if the mesa output format changes, then we have to update the REBOUNDx code. I think the cleanest is if we provide the minimal interface required in REBOUNDx (I.e. simple lists that you can load from text files or however you want) and leave all the rest to the user. 

* That seems strange. The way I would try to debug that is to go into the c code and add something like (pseudocode)

```
if isnan(val){
    fprintf(stderr, "klo=%d\nval[klo]=%f\nval[klo+1]=%f\n", klo, values[klo], values[klo+1])
}
```

to check where you are in the array when things go badly and try to pinpoint it that way. This output will show up in the console window where you launched your jupyter notebook

**Comment by sabaronett** [comment]
Hi Dan,

Thanks for the console output tip on the C-side to help with debugging&mdash;very useful. I started out with your suggestion, `if isnan(val)...`, but didn't notice anything unusual about the outputs except that the NB cell outputted `0.0` that time instead of `nan`. So I tried the code below instead so that I could compare debug outputs between the working MESA reader and the simple text reader vers.:

```c
// interpolation.c
case REBX_INTERPOLATION_SPLINE:
{
    double val = rebx_splint(rebx, interpolator->times, interpolator->values, interpolator->y2, time, &interpolator->klo);       // interpolate at passed time
    fprintf(stderr, "time=%f\nklo=%d\ntimes[klo]=%f\nvalues[klo]=%f\ninterpolate=%f\ntimes[klo+1]=%f\nvalues[klo+1]=%f\n",
        time,
        interpolator->klo,
        interpolator->times[interpolator->klo],
        interpolator->values[interpolator->klo],
        val,
        interpolator->times[interpolator->klo+1],
        interpolator->values[interpolator->klo+1]);
    return val;
}
```

But nothing stands out between the two still (cf. last markdown cell of console outputs in [MesaReader.ipynb](https://github.com/sabaronett/reboundx/blob/c14e45fef4cf1dd18559ad042e15fd7631e6f17d/ipython_examples/MesaReader.ipynb) and [TextReader.ipynb](https://github.com/sabaronett/reboundx/blob/c14e45fef4cf1dd18559ad042e15fd7631e6f17d/ipython_examples/TextReader.ipynb)). I also tried outputs both before and after the calls to `rebx_splint()`, but they were identical.

One thing I did notice that's worth mentioning is that whenever I restart the NB kernel after having run the third cell in either version (when the `Interpolator` instances are constructed), I'll see a random one of the following messages in Jupyter's console:

```bash
`munmap_chunk(): invalid pointer`

`free(): invalid size`

`double free or corruption (out)`

`free(): invalid pointer`
```

Perhaps it's related to the malloc.c issues.

I'll try different type castings and see if that changes anything. I'll also check `py_mesa_reader` source to see what types are being returned.

**Comment by sabaronett** [comment]
@dtamayo,

After spending most of this evening and night, I finally got the version without `mesa_reader` working in 5e65bc37d3fa788d242163ca2b027f06fb88d0a8. Please check out the new [ParameterInterpolation.ipynb](https://github.com/sabaronett/reboundx/blob/5e65bc37d3fa788d242163ca2b027f06fb88d0a8/ipython_examples/ParameterInterpolation.ipynb) and let me know what you think.

I suspect that, when I had "manually" stripped the original MESA output file to isolate the columns (and remove unnecessary rows), I may have inadvertently introduced a hidden escape character (e.g., a Windows CRLF, possibly through whatever editor I was using at the time) that may have gotten picked up during the read and ultimately caused an issue when the `Interpolate` constructor went to initially spline (returning invalid second derivative `y2` values). Long story short, using #24 now ensures that`m.txt` and `r.txt` are valid and working.

In more embarrassing news, I only noticed your pull request #23 a few minutes ago (which explains some misunderstandings on my end, haha). I double-checked my emails and, for whatever reason, I never received a notification. I tried looking into this repo's subscription settings and confirmed I should be receiving subsequent updates on #23, but I can't find anything related to new pull requests in the first place. Anywho, I'll read and go through your request in the morning and hopefully merge tomorrow.

Thanks again for everything.

**Comment by dtamayo** [comment]
Nice detective work! It looks great. I tried merging your branch and running the unit tests but the two related to interpolating backward in time failed. I think once we fix that and update the documentation we're good to merge into master!

**Comment by sabaronett** [comment]
Thanks. Yeah, I just finished reviewing your pull request and should merge it before class in a half hour. After class I'll work on those tests for backwards interpolation and let you know.

By the way, I also just found #18, which I had no idea existed until now, haha. I really need to fix my email notifications. Looks like I ended up on the right track though with switching to pointer arithmetic, but I appreciate the advice you gave.
